pysindy/optimizers/constrained_sr3.py

```diff
@@ -0,0 +1,374 @@
+import warnings
+
+import numpy as np
+from scipy.linalg import cho_factor
+from scipy.linalg import cho_solve
+from scipy.optimize import bisect
+from sklearn.exceptions import ConvergenceWarning
+
+from pysindy.optimizers import BaseOptimizer
+from pysindy.utils import get_prox
+from pysindy.utils import get_reg
+
+
+class ConstrainedSR3(BaseOptimizer):
+    """
+    Sparse relaxed regularized regression.
+
+    Attempts to minimize the objective function
+
+    .. math::
+
+        0.5\\|y-Xw\\|^2_2 + \\lambda \\times R(v)
+        + (0.5 / nu)\\|w-v\\|^2_2
+
+    where :math:`R(v)` is a regularization function. See the following reference
+    for more details:
+
+        Zheng, Peng, et al. "A unified framework for sparse relaxed
+        regularized regression: Sr3." IEEE Access 7 (2018): 1404-1423.
+
+    Parameters
+    ----------
+    threshold : float, optional (default 0.1)
+        Determines the strength of the regularization. When the
+        regularization function R is the l0 norm, the regularization
+        is equivalent to performing hard thresholding, and lambda
+        is chosen to threshold at the value given by this parameter.
+        This is equivalent to choosing lambda = threshold^2 / (2 * nu).
+
+    nu : float, optional (default 1)
+        Determines the level of relaxation. Decreasing nu encourages
+        w and v to be close, whereas increasing nu allows the
+        regularized coefficients v to be farther from w.
+
+    tol : float, optional (default 1e-5)
+        Tolerance used for determining convergence of the optimization
+        algorithm.
+
+    thresholder : string, optional (default 'l0')
+        Regularization function to use. Currently implemented options
+        are 'l0' (l0 norm), 'l1' (l1 norm), and 'cad' (clipped
+        absolute deviation).
+
+    max_iter : int, optional (default 30)
+        Maximum iterations of the optimization algorithm.
+
+    fit_intercept : boolean, optional (default False)
+        Whether to calculate the intercept for this model. If set to false, no
+        intercept will be used in calculations.
+
+    normalize : boolean, optional (default False)
+        This parameter is ignored when fit_intercept is set to False. If True,
+        the regressors X will be normalized before regression by subtracting
+        the mean and dividing by the l2-norm.
+
+    copy_X : boolean, optional (default True)
+        If True, X will be copied; else, it may be overwritten.
+
+    initial_guess : 2D numpy array of floats (default optional)
```

Comment:
Could you add the expected shape for this parameter and update the default value?